File: claude_exp/xv5/shared/ims_connect.py
"""IMS Connect wire protocol. Dual-socket model: to-socket sends requests,
from-socket receives responses."""

from shared.framing import _recv_exact

IRM_HEADER_LEN = 28


def to_ebcdic(s: str, length: int) -> bytes:
    """EBCDIC-encode and left-pad/truncate to exactly `length` bytes."""
    b = s.encode("cp500")
    if len(b) >= length:
        return b[-length:]
    pad = " ".encode("cp500") * (length - len(b))
    return pad + b


PING_TRANSCODE = to_ebcdic("PING0001", 8)


def build_frame(irm_f0, irm_id: bytes, client_id: bytes, mti=None, data: bytes = b"", transcode=None) -> bytes:
    """Build a complete IMS Connect wire frame: 4-byte big-endian length (payload only)
    + 28-byte IMS header + optional TRANS_CODE (8 bytes EBCDIC) + data.
    irm_f0=0x80 -> resume TPIPE (no data). irm_f0=0x00 -> normal request.
    transcode defaults to TRAN+mti when data is present."""
    has_data = len(data) > 0
    if has_data and transcode is None:
        transcode = to_ebcdic(f"TRAN{mti}", 8)

    irm_header = (
        IRM_HEADER_LEN.to_bytes(2, "big")
        + bytes([0x04])
        + bytes([irm_f0])
        + irm_id
        + b"\x00\x00\x00\x00"
        + b"\x00\x15\x10\x01"
        + client_id
    )
    trailer = (transcode + data) if has_data else b""
    payload_len = len(irm_header) + len(trailer)
    return payload_len.to_bytes(4, "big") + irm_header + trailer


def write_response(sock, data: bytes) -> None:
    """Send downstream response: 4-byte big-endian length + data."""
    sock.sendall(len(data).to_bytes(4, "big") + data)


def read_response(sock) -> bytes:
    """Read downstream response. Returns ISO data bytes only (strips length prefix)."""
    len_bytes = _recv_exact(sock, 4)
    length = int.from_bytes(len_bytes, "big")
    return _recv_exact(sock, length)


def read_request(sock) -> tuple:
    """Read IMS Connect request. Returns (irm_f0, client_id_bytes, transcode_bytes, iso_data_bytes)."""
    len_bytes = _recv_exact(sock, 4)
    payload_len = int.from_bytes(len_bytes, "big")
    payload = _recv_exact(sock, payload_len)

    irm_f0 = payload[3]
    client_id = payload[20:28]
    rest = payload[IRM_HEADER_LEN:]
    if rest:
        transcode = rest[:8]
        iso_data = rest[8:]
    else:
        transcode = b""
        iso_data = b""
    return irm_f0, client_id, transcode, iso_data

File: claude_exp/xv6java/src/main/java/com/xv6/shared/ImsConnect.java
package com.xv6.shared;

import java.io.IOException;
import java.net.Socket;
import java.util.Arrays;

/**
 * IMS Connect wire protocol. Dual-socket model: to-socket sends requests, from-socket receives
 * responses. Port of xv5's shared/ims_connect.py.
 */
public final class ImsConnect {

    public static final int IRM_HEADER_LEN = 28;
    public static final byte[] PING_TRANSCODE = toEbcdic("PING0001", 8);

    private ImsConnect() {
    }

    /** EBCDIC-encode and left-pad/truncate to exactly {@code length} bytes. */
    public static byte[] toEbcdic(String s, int length) {
        byte[] b = s.getBytes(Charset500.CHARSET);
        if (b.length >= length) {
            return Arrays.copyOfRange(b, b.length - length, b.length);
        }
        byte[] out = new byte[length];
        byte space = " ".getBytes(Charset500.CHARSET)[0];
        Arrays.fill(out, 0, length - b.length, space);
        System.arraycopy(b, 0, out, length - b.length, b.length);
        return out;
    }

    /**
     * Build a complete IMS Connect wire frame: 4-byte big-endian length (payload only) + 28-byte
     * IMS header + optional TRANS_CODE (8 bytes EBCDIC) + data. irmF0=0x80 -> resume TPIPE (no
     * data). irmF0=0x00 -> normal request. transcode defaults to TRAN+mti when data is present.
     */
    public static byte[] buildFrame(int irmF0, byte[] irmId, byte[] clientId, String mti, byte[] data,
            byte[] transcode) {
        boolean hasData = data.length > 0;
        byte[] resolvedTranscode = transcode;
        if (hasData && resolvedTranscode == null) {
            resolvedTranscode = toEbcdic("TRAN" + mti, 8);
        }

        byte[] irmHeader = new byte[IRM_HEADER_LEN];
        int p = 0;
        irmHeader[p++] = (byte) ((IRM_HEADER_LEN >> 8) & 0xFF);
        irmHeader[p++] = (byte) (IRM_HEADER_LEN & 0xFF);
        irmHeader[p++] = 0x04;
        irmHeader[p++] = (byte) irmF0;
        System.arraycopy(irmId, 0, irmHeader, p, 8);
        p += 8;
        // IRM_NAK_RSNCDE(2) + IRM_RES(2)
        p += 4;
        // IRM_F5, IRM_TIMER, IRM_SOCT, IRM_ES
        irmHeader[p++] = 0x00;
        irmHeader[p++] = 0x15;
        irmHeader[p++] = 0x10;
        irmHeader[p++] = 0x01;
        System.arraycopy(clientId, 0, irmHeader, p, 8);

        byte[] trailer;
        if (hasData) {
            trailer = new byte[resolvedTranscode.length + data.length];
            System.arraycopy(resolvedTranscode, 0, trailer, 0, resolvedTranscode.length);
            System.arraycopy(data, 0, trailer, resolvedTranscode.length, data.length);
        } else {
            trailer = new byte[0];
        }

        int payloadLen = irmHeader.length + trailer.length;
        byte[] out = new byte[4 + payloadLen];
        out[0] = (byte) ((payloadLen >> 24) & 0xFF);
        out[1] = (byte) ((payloadLen >> 16) & 0xFF);
        out[2] = (byte) ((payloadLen >> 8) & 0xFF);
        out[3] = (byte) (payloadLen & 0xFF);
        System.arraycopy(irmHeader, 0, out, 4, irmHeader.length);
        System.arraycopy(trailer, 0, out, 4 + irmHeader.length, trailer.length);
        return out;
    }

    /** Send downstream response: 4-byte big-endian length + data. */
    public static void writeResponse(Socket sock, byte[] data) throws IOException {
        byte[] out = new byte[4 + data.length];
        int len = data.length;
        out[0] = (byte) ((len >> 24) & 0xFF);
        out[1] = (byte) ((len >> 16) & 0xFF);
        out[2] = (byte) ((len >> 8) & 0xFF);
        out[3] = (byte) (len & 0xFF);
        System.arraycopy(data, 0, out, 4, data.length);
        sock.getOutputStream().write(out);
        sock.getOutputStream().flush();
    }

    /** Read downstream response. Returns ISO data bytes only (strips length prefix). */
    public static byte[] readResponse(Socket sock) throws IOException {
        byte[] lenBytes = recvExact(sock, 4);
        int length = bigEndianToInt(lenBytes);
        return recvExact(sock, length);
    }

    public record ImsRequest(int irmF0, byte[] clientId, byte[] transcode, byte[] isoData) {
    }

    /** Read an IMS Connect request. */
    public static ImsRequest readRequest(Socket sock) throws IOException {
        byte[] lenBytes = recvExact(sock, 4);
        int payloadLen = bigEndianToInt(lenBytes);
        byte[] payload = recvExact(sock, payloadLen);

        int irmF0 = payload[3] & 0xFF;
        byte[] clientId = Arrays.copyOfRange(payload, 20, 28);
        byte[] rest = payload.length > IRM_HEADER_LEN
                ? Arrays.copyOfRange(payload, IRM_HEADER_LEN, payload.length)
                : new byte[0];
        byte[] transcode;
        byte[] isoData;
        if (rest.length > 0) {
            transcode = Arrays.copyOfRange(rest, 0, 8);
            isoData = Arrays.copyOfRange(rest, 8, rest.length);
        } else {
            transcode = new byte[0];
            isoData = new byte[0];
        }
        return new ImsRequest(irmF0, clientId, transcode, isoData);
    }

    private static byte[] recvExact(Socket sock, int n) throws IOException {
        byte[] buf = new byte[n];
        int off = 0;
        while (off < n) {
            int read = sock.getInputStream().read(buf, off, n - off);
            if (read < 0) {
                throw new IOException("connection closed while reading");
            }
            off += read;
        }
        return buf;
    }

    private static int bigEndianToInt(byte[] b) {
        return ((b[0] & 0xFF) << 24) | ((b[1] & 0xFF) << 16) | ((b[2] & 0xFF) << 8) | (b[3] & 0xFF);
    }
}
